src/devices/sound/mpeg_audio.cpp

```diff
@@ -719,14 +719,286 @@ void mpeg_audio::retrieve_subbuffer(int step)
 		memcpy(subbuffer[chan], bdata[chan][step], 32*sizeof(subbuffer[0][0]));
 }
 
+const double mpeg_sincos[] = {
+	0.6715589548 /*sin15pi/64*/	, 0.7409511254, /*cos*/
+	0.8032075315 /*cos13pi/64*/	, 0.5956993045, /*sin*/
+	0.8577286100 /*cos11pi/64*/	, 0.5141027442, /*sin*/
+	0.9039892931 /*cos9pi/64*/	, 0.4275550934, /*sin*/
+	0.9415440652 /*cos7pi/64*/	, 0.3368898534, /*sin*/
+	0.2429801799 /*sin5pi/64*/	, 0.9700312532, /*cos*/
+	0.9891765100 /*cos3pi/64*/	, 0.1467304745, /*sin*/
+	0.9987954562 /*cospi/64*/	, 0.0490676743, /*sin*/
+	0.6343932842 /*sin7pi/32*/	, 0.7730104534, /*cos*/
+	0.4713967368 /*sin5pi/32*/	, 0.8819212643, /*cos*/
+	0.2902846773 /*sin3pi/32*/	, 0.9569403357, /*cos*/
+	0.0980171403 /*sinpi/32*/	, 0.9951847267, /*cos*/
+	0.8314696123 /*cos3pi/16*/	, 0.5555702330, /*sin*/
+	0.1950903220 /*sinpi/16*/	, 0.9807852804, /*cos*/
+	0.3826834324 /*sinpi/8*/	, 0.9238795325, /*cos*/
+	0.7071067812 /*sinpi/4*/,
+};
+
+const uint32_t mpeg_sin_15pi_div_64 = 0, mpeg_cos_15pi_div_64 = 1;
+const uint32_t mpeg_cos_13pi_div_64 = 2, mpeg_sin_13pi_div_64 = 3;
+const uint32_t mpeg_cos_11pi_div_64 = 4, mpeg_sin_11pi_div_64 = 5;
+const uint32_t mpeg_cos_9pi_div_64 = 6, mpeg_sin_9pi_div_64 = 7;
+const uint32_t mpeg_cos_7pi_div_64 = 8, mpeg_sin_7pi_div_64 = 9;
+const uint32_t mpeg_sin_5pi_div_64 = 10, mpeg_cos_5pi_div_64 = 11;
+const uint32_t mpeg_cos_3pi_div_64 = 12, mpeg_sin_3pi_div_64 = 13;
+const uint32_t mpeg_cos_pi_div_64 = 14, mpeg_sin_pi_div_64 = 15;
+const uint32_t mpeg_sin_7pi_div_32 = 16, mpeg_cos_7pi_div_32 = 17;
+const uint32_t mpeg_sin_5pi_div_32 = 18, mpeg_cos_5pi_div_32 = 19;
+const uint32_t mpeg_sin_3pi_div_32 = 20, mpeg_cos_3pi_div_32 = 21;
+const uint32_t mpeg_sin_pi_div_32 = 22, mpeg_cos_pi_div_32 = 23;
+const uint32_t mpeg_cos_3pi_div_16 = 24, mpeg_sin_3pi_div_16 = 25;
+const uint32_t mpeg_sin_pi_div_16 = 26, mpeg_cos_pi_div_16 = 27;
+const uint32_t mpeg_sin_pi_div_8 = 28, mpeg_cos_pi_div_8 = 29;
+const uint32_t mpeg_sin_pi_div_4 = 30;
+
 void mpeg_audio::idct32(const double *input, double *output)
 {
-	// Simplest idct32 ever, non-fast at all
-	for(int i=0; i<32; i++) {
-		double s = 0;
-		for(int j=0; j<32; j++)
-			s += input[j] * cos(i*(2*j+1)*M_PI/64);
-		output[i] = s;
+	if (!m_fast_idct32)
+	{
+		// Simplest idct32 ever, non-fast at all
+		for (int i = 0; i < 32; i++) {
+			double s = 0;
+			for (int j = 0; j < 32; j++)
+				s += input[j] * cos(i*(2 * j + 1)*M_PI / 64);
+			output[i] = s;
```

Comment:
Honestly just pre-caching the 32*32 different values of cos in a table would probably make a noticeable speedup for the slow version.

Reply:
let's give that a shot then, bit surprised it wasn't already done if it was likely to make a huge difference

Reply:
and indeed, it is almost as fast